src/extensions/sub-expressions.lisp

```diff
@@ -0,0 +1,49 @@
+(in-package :geb.extensions)
+
+(deftype obj-morph () `(or cat-obj cat-morph t))
+
+(-> common-sub-expressions (obj-morph) (values obj-morph fset:map))
+(defun common-sub-expressions (term)
+  "Compute common sub-expressions and return an object with the
+appropriate sub-expressions uniquely identified"
+  (keep-unique term (compute-common-usages term)))
+
+(-> compute-common-usages (obj-morph) fset:bag)
+(defun compute-common-usages (obj)
+  ;; we are going to be smart about this
+  ;; no manual recursion, we don't need that
+  (labels ((recursive (bag term)
+             (reduce-pointwise (lambda (bag term)
+                                 ;; we don't care about reducing out cat-objs
+                                 (cond ((or (not (typep term 'pointwise-mixin))
+                                            (typep term 'cat-obj))
+                                        bag)
+                                       ;; we don't want to actually add
+                                       ;; to a total that is already
+                                       ;; count
+                                       ((fset:member? term bag)
+                                        (fset:with bag term))
+                                       (t
+                                        (recursive (fset:with bag term) term))))
+                               term
+                               bag)))
+    (values
+     (fset:filter-pairs (lambda (x y) (declare (ignore x)) (<= 2 y))
+                        (recursive (fset:empty-bag) obj)))))
+
+(-> keep-unique (obj-morph fset:bag) (values obj-morph fset:map))
```

Comment:
Is `keep-unique` a function intended for clients who can use it to prevent some object from being factored out as the same as any other?  Or is it only for internal use by the sub-expressions module?  (I don't think there's anything wrong with it either way; I could just use a comment explaining its intent.)

Reply:
Yeah good comment, it's basically the output of `common-sub-expression` where given a bag of repeat expressions, mark the points which are common. Basically the second step in the pass, mostly for internal use

Reply:
Commented